### PR TITLE
[PoC] SEO stuff for products, taxons & related

### PR DESCRIPTION
**Pull request by stloyd**
- [ ] migrations,
- [x] more SEO stuff,
- [x] polish code quality

### CODE REVIEW

**Comment by stloyd** [comment]
I reworked current approach to make it more dynamic, it's not yet that dynamic I think it's good way to extend.

**Comment by pjedrzejewski** [comment]
Looks cool!

**Comment by peteward** [comment]
Looks very good.

SeoExtra looks like it's just a hook/placeholder for now.
I'd be interested to know how you're thinking about fleshing this out?

How exactly are you using the storage as a cache layer?

**Comment by pjedrzejewski** [comment]
Yes, that part is a bit weird?

File: src/Sylius/Component/Taxonomy/Model/TaxonTranslation.php [hunk 1]
```diff
@@ -142,4 +163,58 @@ public function setDescription($description)
 
         return $this;
     }
+
```
Comment (peteward) [inline src/Sylius/Component/Taxonomy/Model/TaxonTranslation.php]: Looking forward to php 5.4 requirement so we can use Traits :)

**Comment by peteward** [comment]
I've been thinking about this...

I'm not sure that the split of Title, Keywords and Description but then having everything else in an 'Extra' field/structure of some sort. I understand why you started with these being explicit and then evolved the design, but I think you should just have a single SeoMetadata field/attribute which can be a serialised array. SEO has come a long way since Title, Keywords and Description, especially with social meta tags which are arguably a lot more useful than Keywords. If you're serious about SEO (and any self-respecting E-commerce business should be) then you'll always want more than the basics.

So rather than having to add these fixed attributes and methods to implement an interface (and you'll probably need two interfaces with your current design) it seems sensible to just have a single SeoMetaInterface with an seoMetadata attribute on each implementing model.


The challenge then is to make the keys (types of metadata) of this seoMetadata dynamic, flexible an default-able. But I think you've done a lot of the really complicated work here already. It looks like we just need a couple of admin interfaces:
* to manage the different possible metadata attributes
* to manage the formulas that you can use to fill them

I think you could also simplify the rendering and make it hierarchical, a bit like symfony forms. So you could do:

```
seo_header_metadata()
```

but also, in case you need the flexibility you could render them separately:

```
{{ seo_metadata('title') }}
{{ seo_metadata('description') }}
```

However, not all SEO attributes are based in the header. IMG ALT content is a vitally important part of SEO. There's also schema.org microdata that is based on mark-up tags rather than the header, so similar to IMG ALT tags.  Perhaps SEO attributes themselves have different characteristics? some can be rendered in a header so could be part of the automatic twig function above, others cannot. 


Finally I'm not sure about the storage/caching layer. Are you trying to save the results of the formula onto the object whenever it changes? If so then what happens if you want to use content/attributes from a relation that changes? Like resource.prototype.name? Will you listen for that too?
It's a bit like the dilemma that we had for search, of trying to magically export content for searching when stuff changes, it quickly gets out of hand as your complexity grows.

But here I think for the sake of micro-optimising you can dynamically generate the content whenever you need it. I'm not sure that saving it is really necessary. You have already pulled the object out of the DB to render it somehow, so presumably all of the details about it that you could use for rendering will be available - and if not, then change the query ;)  You could cache the output for the duration of the request to avoid duplicate rendering of the same IMG ALT tag (for example), but I wouldn't persist it to storage.

If you really want to boost performance with caching in complex situations you need to start thinking of holistic approaches such as a CDN with reverse proxy and ESI - this will boost performance in a much more effective site-wide way without having to introduce coding complexities.

**Comment by stloyd** [comment]
Yes, cache layer is kinda side effect on fighting with Twig & database to avoid as much as possible calling those two services (I even had in mind to pre render and save data in entity ;)).

IMG `alt` attributes IMO are a bit more complicated as they should be created on the Image and just when not found we could create some general formula.

Having array in entity means few problems as they will be quite fragile to what SEO stuff puts into it. I had in mind few possible ways, but I guess that I will start with the simple (without any real backend helpers aka javascript stuff, just simple edit), and then see where it goes.

**Comment by peteward** [comment]
Yeah I think pre-render is a bad idea, although I thought you were doing that, maybe I misread it...

I see what you mean about image alts being a bit more complex, but I don't see this as being too much of a problem. If they are unpersisted objects in their own right, they can just have their alt tags dynamically generated based on the parent model:

```
"{{ resource.product.name }} - {{ resource.product-colour }} in size {{ resource.product.size }}"
```

If they are persisted, such as images for a content page, then they could well have their own SEO attributes related directly to that entity (and you wouldn't want to have Keywords, Title and Description on an Image, so just an SeoMetadata attribute makes sense here).

Why would an array structure be 'fragile'?

**Comment by michalmarcinkowski** [comment]
This will be done in https://github.com/Sylius/Sylius/pull/3350.

**Pull request closed without merging**